### DISCUSSION TITLE
How to set default attributes for metrics ?

### DISCUSSION THREAD

**Discussion by alefcarlos** [open-telemetry/opentelemetry-dotnet#3333] [Q&A] [answered]
Is there a way to add default attribute for all metrics ?

I added attributes to ResourceBuilder:

```csharp
var defaultAttributes = new TagList
{
    { "var1", "value1" },
    { "var2", "value2" },
};

var resourceBuilder = ResourceBuilder
                        .CreateDefault()
                        .AddAttributes(defaultAttributes)
                        .AddService("WebApi1", serviceVersion: assemblyVersion, serviceInstanceId: Environment.MachineName);
```

And then configured OpenTelemetryMetrics to use console exporter to a quick test:

```chsarp
services.AddOpenTelemetryMetrics(options =>
            {
                options.SetResourceBuilder(resourceBuilder);
                options.AddMeter("Super.WebApi");
                options.AddConsoleExporter();
            });
```
And at Controller:

```csharp
var attributes = new TagList
{
    { "my_label", "value1" },
};
_metrics.TransactionAuthorized.Add(1, attributes);
```

I'd expect to see `my_label`, `var1` and `var2` on Console, but only `my_label` was shown:

```
Export TransactionAuthorized, Meter: Super.WebApi/1.0.0
(2022-06-02T19:32:47.5247319Z, 2022-06-02T19:34:26.6444907Z] my_label: value1 LongSum
Value: 7
```

Is there anything I could do ?

**Comment by cijothomas** [comment, marked as answer]
Resource is a way to model this. (The console exporter for metrics is ignoring resource - its a bug. will send a fix).
But you must be using something other than Console for your actual service right? Like an OTLP exporter. it should already support Resources.

**Reply by alefcarlos** [reply]
I'm using OTLP exporter with `dynatrace` exporter for metrics, but the same problem occurs :(

**Reply by cijothomas** [reply]
Can you check if OTLPExporter works fine? (not the dynatrace backend)
https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/examples/Console/TestMetrics.cs You can use this example to try the OTLP Exporter

**Reply by alefcarlos** [reply]
I already have an implementation for OTLPExporter:

```
          services.AddOpenTelemetryMetrics(options =>
          {
              var metricsExporter = Configuration.GetValue<string>("UseMetricsExporter").ToLowerInvariant();

              options.SetResourceBuilder(resourceBuilder);
              options.AddMeter("Super.WebApi");

              switch (metricsExporter)
              {
                  case "prometheus":
                      options.AddPrometheusExporter();
                      break;

                  case "otlp":
                      options.AddOtlpExporter(otlpOptions =>
                      {
                          otlpOptions.Endpoint = new Uri(Configuration.GetValue<string>("Otlp:Endpoint"));
                      });
                      break;

                  case "console":
                      options.AddConsoleExporter();
                      break;

                  default:
                      break;
              };
          });
```

But where can I view the matrics data? Prometheus sounds good ?

**Reply by alefcarlos** [reply]
I tried OtlpExporter + Prometheus:

![image](https://user-images.githubusercontent.com/5452336/171733937-87f9c223-305b-454f-988e-7e48918024ef.png)

Only `my_label` was ingested.

**Reply by cijothomas** [reply]
> But where can I view the matrics data

The example I shared has OTel Collecting with logging output enabled - so it output metrics to console, which should show resources.

(Prometheus does not support resource yet, its pending spec.)

**Reply by alefcarlos** [reply]
I get it. I can see the resources attributes.

![image](https://user-images.githubusercontent.com/5452336/171767781-2d83d2ad-434c-4f40-a36e-826e42fbae5a.png)

I should open an issue/discussion on Dynatrace exporter to understand that.

Thank you so much, @cijothomas  !

**Reply by xavirec** [reply]
Is there any plan to implement this in the PrometheusExporter?

**Reply by cijothomas** [reply]
> Is there any plan to implement this in the PrometheusExporter?

Its a spec question. See experimental version here : https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/datamodel.md#prometheus-compatibility
